src/energyhub.py

```diff
@@ -735,136 +754,247 @@ def _write_solution_diagnostics(self, save_path):
-                            .periods[period]
-                            .node_blocks[node]
-                            .tech_blocks_active
+                            model.periods[period].node_blocks[node].tech_blocks_active
                         ):
-                            self._monte_carlo_technologies(node, tec)
+                            self._monte_carlo_technologies(period, node, tec)
 
             if "Networks" in config["optimization"]["monte_carlo"]["on_what"]["value"]:
-                for netw in model.network_block:
-                    self._monte_carlo_networks(netw)
-
-            if (
-                "ImportPrices"
-                in config["optimization"]["monte_carlo"]["on_what"]["value"]
-            ):
-                for node in model.node_blocks:
-                    for car in model.node_blocks[node].set_carriers:
-                        self._monte_carlo_import_prices(node, car)
-
-            if (
-                "ExportPrices"
-                in config["optimization"]["monte_carlo"]["on_what"]["value"]
-            ):
-                for node in model.node_blocks:
-                    for car in model.node_blocks[node].set_carriers:
-                        self._monte_carlo_export_prices(node, car)
-
-    def _monte_carlo_technologies(self, node, tec):
+                for period in model.periods:
+                    for netw in model.periods[period].network_block:
+                        self._monte_carlo_networks(period, netw)
+
+            if "Import" in config["optimization"]["monte_carlo"]["on_what"]["value"]:
+                for period in model.periods:
+                    for node in model.periods[period].node_blocks:
+                        for car in model.periods[period].node_blocks[node].set_carriers:
+                            self._monte_carlo_import_prices(period, node, car)
+
+            if "Export" in config["optimization"]["monte_carlo"]["on_what"]["value"]:
+                for period in model.periods:
+                    for node in model.periods[period].node_blocks:
+                        for car in model.periods[period].node_blocks[node].set_carriers:
+                            self._monte_carlo_export_prices(period, node, car)
+
+        elif config["optimization"]["monte_carlo"]["type"]["value"] == 2:
+            MC_parameters = self.data.monte_carlo_specs
+
+            for index, row in MC_parameters.iterrows():
+                if row["type"] == "Technologies":
+                    MC_technology_row = row
+                    tec = MC_technology_row["name"]
+
+                    for period in model.periods:
+                        for node in model.periods[period].node_blocks:
+                            if tec in (
+                                model.periods[period]
+                                .node_blocks[node]
+                                .tech_blocks_active
+                            ):
+                                if (
```

Comment:
Can we reformulate this: 

if capex_model == 1:
technology_parameters = MC_technology_rows

if capex_model == 3:
technology_parameters = []
...
technology_parameters.append(row)
else:
warning(other capex models not defined)

outside of if:
self._monte_carlo_technologies(..., technology_parameters)

That way it becomes more clear what you are doing with the technology_parameters for each capex model and that you call _monte_carlo_technologies only once per tec.